lib/ansible/modules/cloud/amazon/ec2_lc.py

```diff
@@ -181,169 +202,224 @@ def create_block_device(module, volume):
-    bdm = BlockDeviceMapping()
+    security_groups = module.params.get('security_groups')
+
+    bdm = {}
+
+    convert_list = ['image_id', 'instance_type', 'instance_type', 'instance_id', 'placement_tenancy', 'key_name', 'kernel_id', 'ramdisk_id',
+                    'instance_profile_name', 'spot_price']
+
+    launch_config = (snake_dict_to_camel_dict(dict((k.capitalize(), str(v)) for k, v in module.params.items() if v is not None and k in convert_list)))
 
     if user_data_path:
         try:
             with open(user_data_path, 'r') as user_data_file:
                 user_data = user_data_file.read()
         except IOError as e:
-            module.fail_json(msg=str(e), exception=traceback.format_exc())
+            module.fail_json(msg="Failed to open file for reading", exception=traceback.format_exc())
 
     if volumes:
         for volume in volumes:
             if 'device_name' not in volume:
                 module.fail_json(msg='Device name must be set for volume')
-            # Minimum volume size is 1GB. We'll use volume size explicitly set to 0
-            # to be a signal not to create this volume
+            # Minimum volume size is 1GB. We'll use volume size explicitly set to 0 to be a signal not to create this volume
             if 'volume_size' not in volume or int(volume['volume_size']) > 0:
-                bdm[volume['device_name']] = create_block_device(module, volume)
-
-    lc = LaunchConfiguration(
-        name=name,
-        image_id=image_id,
-        key_name=key_name,
-        security_groups=security_groups,
-        user_data=user_data,
-        block_device_mappings=[bdm],
-        instance_type=instance_type,
-        kernel_id=kernel_id,
-        spot_price=spot_price,
-        instance_monitoring=instance_monitoring,
-        associate_public_ip_address=assign_public_ip,
-        ramdisk_id=ramdisk_id,
-        instance_profile_name=instance_profile_name,
-        ebs_optimized=ebs_optimized,
-        classic_link_vpc_security_groups=classic_link_vpc_security_groups,
-        classic_link_vpc_id=classic_link_vpc_id,
-    )
+                bdm.update(create_block_device_meta(module, volume))
+
+    try:
+        launch_configs = connection.describe_launch_configurations(LaunchConfigurationNames=[name]).get('LaunchConfigurations')
+    except botocore.exceptions.ClientError as e:
+        module.fail_json(msg="Failed to describe launch configuration by name", exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
 
-    launch_configs = connection.get_all_launch_configurations(names=[name])
     changed = False
-    if not launch_configs:
+    result = {}
+
+    launch_config['LaunchConfigurationName'] = name
+
+    if security_groups is not None:
+        launch_config['SecurityGroups'] = security_groups
+
+    if classic_link_vpc_id is not None:
+        launch_config['ClassicLinkVPCId'] = classic_link_vpc_id
+
+    if instance_monitoring:
```

Comment:
It looks like just setting `instance_monitoring` to `true` now won't really do anything, since basic monitoring is on by default and the value of `advanced_instance_monitoring` would default to `False`. This seems confusing.